File: sdk/node/src/retry.ts
export function computeBackoff(attempt: number, base: number, cap: number): number {
  if (attempt < 1) {
    throw new RangeError("attempt must be >= 1");
  }
  return Math.min(cap, base * 2 ** (attempt - 1));
}

export function parseRetryAfter(value: string | null, now: Date): number | null {
  if (value === null) {
    return null;
  }
  const trimmed = value.trim();
  if (trimmed.length === 0) {
    return null;
  }
  if (/^-?\d+(?:\.\d+)?$/.test(trimmed)) {
    const asNumber = Number(trimmed);
    return Math.max(0, asNumber);
  }
  const asDate = new Date(trimmed);
  if (Number.isNaN(asDate.getTime())) {
    return null;
  }
  const deltaSeconds = (asDate.getTime() - now.getTime()) / 1000;
  return Math.max(0, deltaSeconds);
}

File: sdk/python/src/sentinel/_retry.py
from __future__ import annotations

from datetime import datetime
from email.utils import parsedate_to_datetime


def compute_backoff(*, attempt: int, base: float, cap: float) -> float:
    if attempt < 1:
        raise ValueError("attempt must be >= 1")
    return float(min(cap, base * (2 ** (attempt - 1))))


def parse_retry_after(value: str | None, *, now: datetime) -> float | None:
    if value is None:
        return None
    value = value.strip()
    if not value:
        return None
    try:
        seconds = float(value)
    except ValueError:
        seconds = None
    if seconds is not None:
        return max(0.0, seconds)
    try:
        target = parsedate_to_datetime(value)
    except (TypeError, ValueError):
        return None
    if target is None:
        return None
    delta = (target - now).total_seconds()
    return max(0.0, delta)
